src/gui/SubWindow.cpp

```diff
@@ -26,42 +26,245 @@
+
+    m_restoreBtn = new QPushButton( embed::getIconPixmap( "restore" ), QString::null, this );
+    m_restoreBtn->resize(17, 17);
+    m_restoreBtn->setFocusPolicy(Qt::NoFocus);
+    m_restoreBtn->setToolTip(tr("Restore"));
+    connect(m_restoreBtn, SIGNAL(clicked(bool)),this,SLOT(showNormal()));
+
+    // QLabel for window title and shadow effect
+    m_shadow = new QGraphicsDropShadowEffect();
+    m_shadow->setColor(m_textShadowColor);
+    m_shadow->setBlurRadius(3);
+    m_shadow->setXOffset(1);
+    m_shadow->setYOffset(1);
+
+    m_windowTitle = new QLabel(this);
+    m_windowTitle->setFocusPolicy(Qt::NoFocus);
+    m_windowTitle->setGraphicsEffect(m_shadow);
+}
+
+
+
+
+void SubWindow::paintEvent(QPaintEvent *)
+{
+    QPainter p(this);
+    const int titleBarHeight = 24;
+    QRect rect(0, 0, width(), titleBarHeight);
+    bool isActive = SubWindow::mdiArea()->activeSubWindow() == this;
+
+    p.fillRect(rect, isActive ? activeColor() : p.pen().brush() );
+
+    // window border
+    p.setPen(borderColor());
+
+    // bottom, left, and right lines
+    p.drawLine( 0, height() - 1, width(), height() - 1 );
+    p.drawLine( 0, titleBarHeight, 0, height() - 1 );
+    p.drawLine( width() - 1, titleBarHeight, width() - 1, height() - 1 );
+
+    //window icon
+    QPixmap winicon(widget()->windowIcon().pixmap(QSize(17,17)));
+    p.drawPixmap(3,3,17,17,winicon);
 }
 
+
+
+
 QRect SubWindow::getTrueNormalGeometry() const
 {
 	return m_trackedNormalGeom;
 }
 
+
+
+
+QBrush SubWindow::activeColor() const
+{
+    return m_activeColor;
+}
+
+
+
+
+QColor SubWindow::textShadowColor() const
+{
+    return m_textShadowColor;
+}
+
+
+
+
+QColor SubWindow::borderColor() const
+{
+	return m_borderColor;
+}
+
+
+
+
+void SubWindow::setActiveColor( const QBrush & b )
+{
+    m_activeColor = b;
+}
+
+
+
+
+void SubWindow::setTextShadowColor(const QColor & c)
+{
+    m_textShadowColor = c;
+}
+
+
+
+
+void SubWindow::setBorderColor(const QColor &c)
+{
+	m_borderColor = c;
+}
+
+
+
+
 void SubWindow::moveEvent(QMoveEvent * event)
 {
 	QMdiSubWindow::moveEvent(event);
 	// if the window was moved and ISN'T minimized/maximized/fullscreen,
 	//   then save the current position
-	if (!isMaximized() && !isMinimized() && !isFullScreen())
+	if(!isMaximized() && !isMinimized() && !isFullScreen())
 	{
 		m_trackedNormalGeom.moveTopLeft(event->pos());
 	}
 }
 
+
+
+
 void SubWindow::resizeEvent(QResizeEvent * event)
 {
+    /* button adjustments*/
+    m_closeBtn->move(width()-22,3);
+    m_maximizeBtn->hide();
+    m_restoreBtn->hide();
+    int buttonBarWidth = 22;
+    if( windowFlags() & Qt::WindowMinimizeButtonHint)
+    {
+        int d = 0;
+        buttonBarWidth = buttonBarWidth + 18;
+        if ( windowFlags() & Qt::WindowMaximizeButtonHint)
+        {
+            d = 18;
+            buttonBarWidth = buttonBarWidth + 18;
+            m_maximizeBtn->show();
+            m_maximizeBtn->move(width()-40,3);
```

Comment:
why is it `width-40`, and why can't a variable handle this? Generally this whole chunk of code looks too complex and is full of magic numbers.